File: mac/VoiceTests/DictationMachineTests.swift
import XCTest
@testable import Voice

final class DictationMachineTests: XCTestCase {
    private func ready() -> DictationMachine { var m = DictationMachine(); _ = m.handle(.modelReady); return m }

    func testIgnoresHotkeyWhileModelLoading() {
        var m = DictationMachine()
        XCTAssertEqual(m.handle(.hotkeyDown(nil)), [.hud(.modelLoading(0))])
        XCTAssertTrue(m.queue.isEmpty)
    }

    /// A double-tap during the first launch's model load must not start a latched session.
    ///
    /// The gesture itself is valid — `TapLatch` says `.latch`, as asserted below — so nothing in the
    /// key path stops it. What stops it is `canLatch`: `.hotkeyDown` refuses to start a dictation while
    /// the model is loading, and latching over a dictation that never began leaves the bar claiming a
    /// live microphone and makes the next press end a session that does not exist. Windows guarded this
    /// in `Coordinator.Apply` (no latch without a capture); the Mac's key path had not.
    func testADoubleTapWhileTheModelIsLoadingCannotLatch() {
        var m = DictationMachine()
        XCTAssertFalse(m.canLatch)
        XCTAssertEqual(m.handle(.hotkeyDown(nil)), [.hud(.modelLoading(0))])
        XCTAssertTrue(m.queue.isEmpty)

        // The gesture does reach `.latch`, so the refusal above is the machine's and not an accident
        // of the tap timing.
        let t0 = Date(timeIntervalSince1970: 1_000_000)
        var latch = TapLatch()
        XCTAssertEqual(latch.handle(.press, at: t0), [.startDictation])
        XCTAssertEqual(latch.handle(.release, at: t0.addingTimeInterval(0.1)), [.holdOpen])
        XCTAssertEqual(latch.handle(.press, at: t0.addingTimeInterval(0.2)), [.latch])

        _ = m.handle(.modelReady)
        XCTAssertTrue(m.canLatch)
        XCTAssertEqual(m.handle(.hotkeyDown(nil)), [.startRecording, .hud(.listening)])
    }

    func testHappyPath() {
        var m = ready()
        XCTAssertEqual(m.handle(.hotkeyDown(nil)), [.startRecording, .hud(.listening)])
        XCTAssertEqual(m.handle(.hotkeyUp), [.stopRecording])
        let id = m.queue[0].clientId
        XCTAssertEqual(m.handle(.audioStopped(samples: [0.1, 0.2], ms: 1200, speech: true)), [.transcribe(id), .hud(.transcribing(progress: nil))])
        XCTAssertEqual(m.handle(.transcribed(id, text: "olá", language: "pt", ms: 900)), [.refine(id), .hud(.cleaning)])
        XCTAssertEqual(m.handle(.refined(id, .cleaned("Olá."))), [.insert(id, "Olá.")])
        XCTAssertEqual(m.handle(.inserted(id, .cleaned)), [.reportInjected(id, .cleaned), .hud(.done(preview: "Olá.", via: Strings.viaCleaned))])
        XCTAssertTrue(m.queue.isEmpty)
    }

    func testShortOrSilentRecordingIsDropped() {
        var m = ready()
        _ = m.handle(.hotkeyDown(nil)); _ = m.handle(.hotkeyUp)
        XCTAssertEqual(m.handle(.audioStopped(samples: [], ms: 200, speech: true)), [.hud(.message(Strings.nothingHeard))])
        XCTAssertTrue(m.queue.isEmpty)
        _ = m.handle(.hotkeyDown(nil)); _ = m.handle(.hotkeyUp)
        XCTAssertEqual(m.handle(.audioStopped(samples: [0], ms: 3000, speech: false)), [.hud(.message(Strings.nothingHeard))])
    }

    func testCancelWhileListeningDiscards() {
        var m = ready()
        _ = m.handle(.hotkeyDown(nil))
        XCTAssertEqual(m.handle(.cancelRequested), [.discardRecording, .hud(.hidden)])
        XCTAssertTrue(m.queue.isEmpty)
        // a hotkeyUp after cancel is a no-op
        XCTAssertEqual(m.handle(.hotkeyUp), [])
    }

    func testBurstKeepsPasteOrder() {
        var m = ready()
        _ = m.handle(.hotkeyDown(nil)); _ = m.handle(.hotkeyUp)
        let a = m.queue[0].clientId
        _ = m.handle(.audioStopped(samples: [1], ms: 1000, speech: true))
        _ = m.handle(.transcribed(a, text: "a", language: "en", ms: 1))
        // second dictation starts while the first is refining
        XCTAssertEqual(m.handle(.hotkeyDown(nil)), [.startRecording, .hud(.listening)])
        _ = m.handle(.hotkeyUp)
        let b = m.queue[1].clientId
        _ = m.handle(.audioStopped(samples: [1], ms: 1000, speech: true))
        _ = m.handle(.transcribed(b, text: "b", language: "en", ms: 1))
        // b's cleanup arrives first: it must wait
        XCTAssertEqual(m.handle(.refined(b, .cleaned("B."))), [])
        XCTAssertEqual(m.handle(.refined(a, .cleaned("A."))), [.insert(a, "A.")])
        XCTAssertEqual(m.handle(.inserted(a, .cleaned)), [.reportInjected(a, .cleaned), .insert(b, "B.")])
        XCTAssertEqual(m.handle(.inserted(b, .cleaned)), [.reportInjected(b, .cleaned), .hud(.done(preview: "B.", via: Strings.viaCleaned))])
    }

    func testRawFallbackPastesRawAndReportsRaw() {
        var m = ready()
        _ = m.handle(.hotkeyDown(nil)); _ = m.handle(.hotkeyUp)
        let id = m.queue[0].clientId
        _ = m.handle(.audioStopped(samples: [1], ms: 1000, speech: true))
        _ = m.handle(.transcribed(id, text: "raw text", language: "en", ms: 1))
        XCTAssertEqual(m.handle(.refined(id, .rawFallback(.offline))), [.insert(id, "raw text")])
        // Said after the paste, in place of `.done`: said before it, `.done` replaced it within milliseconds.
        XCTAssertEqual(m.handle(.inserted(id, .raw)), [.reportInjected(id, .raw), .hud(.message(Strings.pastedRaw))])
    }

    func testAnInvalidTokenSaysSoOnceTheRawTextIsPasted() {
        var m = ready()
        _ = m.handle(.hotkeyDown(nil)); _ = m.handle(.hotkeyUp)
        let id = m.queue[0].clientId
        _ = m.handle(.audioStopped(samples: [1], ms: 1000, speech: true))
        _ = m.handle(.transcribed(id, text: "raw text", language: "en", ms: 1))
        XCTAssertEqual(m.handle(.refined(id, .rawFallback(.unauthorized))), [.insert(id, "raw text")])
        XCTAssertEqual(m.handle(.inserted(id, .raw)), [.reportInjected(id, .raw), .hud(.message(Strings.tokenInvalid))])
    }

    func testTranscriptionFailureDropsTheDictation() {
        var m = ready()
        _ = m.handle(.hotkeyDown(nil)); _ = m.handle(.hotkeyUp)
        let id = m.queue[0].clientId
        _ = m.handle(.audioStopped(samples: [1], ms: 1000, speech: true))
        XCTAssertEqual(m.handle(.transcriptionFailed(id)), [.hud(.message(Strings.nothingHeard))])
        XCTAssertTrue(m.queue.isEmpty)
    }
}

File: windows/Spit.Core.Tests/DictationMachineTests.cs
namespace Spit.Core.Tests;

// Port of mac/VoiceTests/DictationMachineTests.swift.
public sealed class DictationMachineTests
{
    private static DictationMachine Ready()
    {
        var m = new DictationMachine();
        _ = m.Handle(new MachineEvent.ModelReady());
        return m;
    }

    private static Effect[] Fx(params Effect[] effects) => effects;
    private static Effect Hud(HUDState state) => new Effect.Hud(state);

    [Fact]
    public void testIgnoresHotkeyWhileModelLoading()
    {
        var m = new DictationMachine();
        Assert.Equal(Fx(Hud(new HUDState.ModelLoading(0))), m.Handle(new MachineEvent.HotkeyDown(null)));
        Assert.Empty(m.Queue);
    }

    /// A double-tap during the first launch's model load must not start a latched session. The gesture itself
    /// is valid — `TapLatch` says `Latch` — so what stops it is `CanLatch`.
    [Fact]
    public void testADoubleTapWhileTheModelIsLoadingCannotLatch()
    {
        var m = new DictationMachine();
        Assert.False(m.CanLatch);
        Assert.Equal(Fx(Hud(new HUDState.ModelLoading(0))), m.Handle(new MachineEvent.HotkeyDown(null)));
        Assert.Empty(m.Queue);

        // The gesture does reach `Latch`, so the refusal above is the machine's and not an accident of the tap timing.
        var t0 = DateTimeOffset.FromUnixTimeSeconds(1_000_000);
        var latch = new TapLatch();
        Assert.Equal([TapLatch.Outcome.StartDictation], latch.Handle(HotkeyAction.Press, t0));
        Assert.Equal([TapLatch.Outcome.HoldOpen], latch.Handle(HotkeyAction.Release, t0.AddMilliseconds(100)));
        Assert.Equal([TapLatch.Outcome.Latch], latch.Handle(HotkeyAction.Press, t0.AddMilliseconds(200)));

        _ = m.Handle(new MachineEvent.ModelReady());
        Assert.True(m.CanLatch);
        Assert.Equal(Fx(new Effect.StartRecording(), Hud(new HUDState.Listening())), m.Handle(new MachineEvent.HotkeyDown(null)));
    }

    [Fact]
    public void testHappyPath()
    {
        var m = Ready();
        Assert.Equal(Fx(new Effect.StartRecording(), Hud(new HUDState.Listening())), m.Handle(new MachineEvent.HotkeyDown(null)));
        Assert.Equal(Fx(new Effect.StopRecording()), m.Handle(new MachineEvent.HotkeyUp()));
        var id = m.Queue[0].ClientId;
        Assert.Equal(Fx(new Effect.Transcribe(id), Hud(new HUDState.Transcribing(null))),
            m.Handle(new MachineEvent.AudioStopped([0.1f, 0.2f], 1200, true)));
        Assert.Equal(Fx(new Effect.Refine(id), Hud(new HUDState.Cleaning())),
            m.Handle(new MachineEvent.Transcribed(id, "olá", "pt", 900)));
        Assert.Equal(Fx(new Effect.Insert(id, "Olá.")), m.Handle(new MachineEvent.Refined(id, new RefineResult.Cleaned("Olá."))));
        Assert.Equal(Fx(new Effect.ReportInjected(id, Injected.Cleaned), Hud(new HUDState.Done("Olá.", Strings.ViaCleaned))),
            m.Handle(new MachineEvent.Inserted(id, Injected.Cleaned)));
        Assert.Empty(m.Queue);
    }

    [Fact]
    public void testShortOrSilentRecordingIsDropped()
    {
        var m = Ready();
        _ = m.Handle(new MachineEvent.HotkeyDown(null)); _ = m.Handle(new MachineEvent.HotkeyUp());
        Assert.Equal(Fx(Hud(new HUDState.Message(Strings.NothingHeard))), m.Handle(new MachineEvent.AudioStopped([], 200, true)));
        Assert.Empty(m.Queue);
        _ = m.Handle(new MachineEvent.HotkeyDown(null)); _ = m.Handle(new MachineEvent.HotkeyUp());
        Assert.Equal(Fx(Hud(new HUDState.Message(Strings.NothingHeard))), m.Handle(new MachineEvent.AudioStopped([0], 3000, false)));
    }

    [Fact]
    public void testCancelWhileListeningDiscards()
    {
        var m = Ready();
        _ = m.Handle(new MachineEvent.HotkeyDown(null));
        Assert.Equal(Fx(new Effect.DiscardRecording(), Hud(new HUDState.Hidden())), m.Handle(new MachineEvent.CancelRequested()));
        Assert.Empty(m.Queue);
        // a hotkeyUp after cancel is a no-op
        Assert.Empty(m.Handle(new MachineEvent.HotkeyUp()));
    }

    [Fact]
    public void testBurstKeepsPasteOrder()
    {
        var m = Ready();
        _ = m.Handle(new MachineEvent.HotkeyDown(null)); _ = m.Handle(new MachineEvent.HotkeyUp());
        var a = m.Queue[0].ClientId;
        _ = m.Handle(new MachineEvent.AudioStopped([1], 1000, true));
        _ = m.Handle(new MachineEvent.Transcribed(a, "a", "en", 1));
        // second dictation starts while the first is refining
        Assert.Equal(Fx(new Effect.StartRecording(), Hud(new HUDState.Listening())), m.Handle(new MachineEvent.HotkeyDown(null)));
        _ = m.Handle(new MachineEvent.HotkeyUp());
        var b = m.Queue[1].ClientId;
        _ = m.Handle(new MachineEvent.AudioStopped([1], 1000, true));
        _ = m.Handle(new MachineEvent.Transcribed(b, "b", "en", 1));
        // b's cleanup arrives first: it must wait
        Assert.Empty(m.Handle(new MachineEvent.Refined(b, new RefineResult.Cleaned("B."))));
        Assert.Equal(Fx(new Effect.Insert(a, "A.")), m.Handle(new MachineEvent.Refined(a, new RefineResult.Cleaned("A."))));
        Assert.Equal(Fx(new Effect.ReportInjected(a, Injected.Cleaned), new Effect.Insert(b, "B.")),
            m.Handle(new MachineEvent.Inserted(a, Injected.Cleaned)));
        Assert.Equal(Fx(new Effect.ReportInjected(b, Injected.Cleaned), Hud(new HUDState.Done("B.", Strings.ViaCleaned))),
            m.Handle(new MachineEvent.Inserted(b, Injected.Cleaned)));
    }

    [Fact]
    public void testRawFallbackPastesRawAndReportsRaw()
    {
        var m = Ready();
        _ = m.Handle(new MachineEvent.HotkeyDown(null)); _ = m.Handle(new MachineEvent.HotkeyUp());
        var id = m.Queue[0].ClientId;
        _ = m.Handle(new MachineEvent.AudioStopped([1], 1000, true));
        _ = m.Handle(new MachineEvent.Transcribed(id, "raw text", "en", 1));
        Assert.Equal(Fx(new Effect.Insert(id, "raw text")),
            m.Handle(new MachineEvent.Refined(id, new RefineResult.RawFallback(FallbackReason.Offline))));
        // Said after the paste, in place of Done: said before it, Done replaced it within milliseconds.
        Assert.Equal(Fx(new Effect.ReportInjected(id, Injected.Raw), Hud(new HUDState.Message(Strings.PastedRaw))),
            m.Handle(new MachineEvent.Inserted(id, Injected.Raw)));
    }

    [Fact]
    public void testAnInvalidTokenSaysSoOnceTheRawTextIsPasted()
    {
        var m = Ready();
        _ = m.Handle(new MachineEvent.HotkeyDown(null)); _ = m.Handle(new MachineEvent.HotkeyUp());
        var id = m.Queue[0].ClientId;
        _ = m.Handle(new MachineEvent.AudioStopped([1], 1000, true));
        _ = m.Handle(new MachineEvent.Transcribed(id, "raw text", "en", 1));
        Assert.Equal(Fx(new Effect.Insert(id, "raw text")),
            m.Handle(new MachineEvent.Refined(id, new RefineResult.RawFallback(FallbackReason.Unauthorized))));
        Assert.Equal(Fx(new Effect.ReportInjected(id, Injected.Raw), Hud(new HUDState.Message(Strings.TokenInvalid))),
            m.Handle(new MachineEvent.Inserted(id, Injected.Raw)));
    }

    [Fact]
    public void testTranscriptionFailureDropsTheDictation()
    {
        var m = Ready();
        _ = m.Handle(new MachineEvent.HotkeyDown(null)); _ = m.Handle(new MachineEvent.HotkeyUp());
        var id = m.Queue[0].ClientId;
        _ = m.Handle(new MachineEvent.AudioStopped([1], 1000, true));
        Assert.Equal(Fx(Hud(new HUDState.Message(Strings.NothingHeard))), m.Handle(new MachineEvent.TranscriptionFailed(id)));
        Assert.Empty(m.Queue);
    }
}
